File: Sources/Usage/GeminiUsageFetcher.swift
import Foundation

/// Fetches the Gemini CLI's Code Assist quota buckets, authenticating with
/// `~/.gemini/oauth_creds.json` and refreshing through Google's token
/// endpoint when the access token is expired or rejected. Display-only —
/// no transcript monitoring, no alarms live here.
enum GeminiUsageFetcher {
    enum Outcome {
        case success(GeminiQuotaSnapshot)
        /// Refresh path exhausted — the user has to run `gemini` and re-auth.
        case reauthRequired
        /// A refresh was required but no client id/secret could be extracted
        /// from a local gemini-cli install (and no env override).
        case needsCLIInstall
        /// Google's consumer-tier shutdown verdict — the account moved to
        /// Antigravity. A state, not an error.
        case migratedToAntigravity
        case failed(String)
        /// settings.json declares api-key / vertex-ai.
        case unsupportedAuth(String)
        /// No oauth_creds.json — Gemini CLI never signed in on this machine.
        case notInstalled
    }

    private static let loadCodeAssistURL =
        "https://cloudcode-pa.googleapis.com/v1internal:loadCodeAssist"
    private static let quotaURL =
        "https://cloudcode-pa.googleapis.com/v1internal:retrieveUserQuota"
    private static let tokenURL = "https://oauth2.googleapis.com/token"

    static func fetch() async -> Outcome {
        switch GeminiCredentials.detect() {
        case .notInstalled: return .notInstalled
        case .unsupportedAuth(let type): return .unsupportedAuth(type)
        case .oauthPersonal: break
        }
        let credsURL = GeminiCredentials.credsURL()
        guard var creds = GeminiCredentials.loadCreds(from: credsURL) else {
            return .notInstalled
        }

        // Proactive refresh near expiry (Google access tokens live ~1h).
        if GeminiCredentials.needsRefresh(creds), creds.refreshToken != nil {
            switch await refresh(creds: creds, credsURL: credsURL) {
            case .refreshed(let fresh): creds = fresh
            case .noClientCredentials: return .needsCLIInstall
            // Clock skew can make a working token look expired — still try.
            case .failed: break
            }
        }

        switch await fetchSnapshot(token: creds.accessToken) {
        case .success(let snapshot):
            return .success(snapshot)
        case .migrated:
            return .migratedToAntigravity
        case .unauthorized:
            guard creds.refreshToken != nil else { return .reauthRequired }
            switch await refresh(creds: creds, credsURL: credsURL) {
            case .noClientCredentials: return .needsCLIInstall
            case .failed: return .reauthRequired
            case .refreshed(let fresh):
                switch await fetchSnapshot(token: fresh.accessToken) {
                case .success(let snapshot): return .success(snapshot)
                case .migrated:              return .migratedToAntigravity
                case .unauthorized:          return .reauthRequired
                case .failed(let message):   return .failed(message)
                }
            }
        case .failed(let message):
            return .failed(message)
        }
    }

    // MARK: - Quota

    private enum SnapshotResult {
        case success(GeminiQuotaSnapshot)
        case migrated
        case unauthorized
        case failed(String)
    }

    private static func fetchSnapshot(token: String) async -> SnapshotResult {
        let profileOutcome = await post(
            url: loadCodeAssistURL,
            token: token,
            body: ["metadata": ["ideType": "GEMINI_CLI", "pluginType": "GEMINI"]]
        )

        var profile: GeminiQuotaParser.CodeAssistProfile?
        switch profileOutcome {
        case .http(401, _), .http(403, _):
            return .unauthorized
        case .http(let status, let data):
            if GeminiQuotaParser.isMigrationSignal(data) { return .migrated }
            // The tier/project call is garnish for the quota call — a non-200
            // here still lets the bucket fetch try with an empty project.
            if status == 200 { profile = GeminiQuotaParser.parseLoadCodeAssist(data) }
        case .transport(let message):
            return .failed(message)
        }

        var quotaBody: [String: Any] = [:]
        if let projectID = profile?.projectID { quotaBody["project"] = projectID }
        switch await post(url: quotaURL, token: token, body: quotaBody) {
        case .http(401, _), .http(403, _):
            return .unauthorized
        case .http(let status, let data):
            if GeminiQuotaParser.isMigrationSignal(data) { return .migrated }
            guard status == 200 else { return .failed("http \(status)") }
            guard let buckets = GeminiQuotaParser.parseQuota(data) else {
                return .failed("parse error")
            }
            return .success(GeminiQuotaSnapshot(
                buckets: buckets,
                tierID: profile?.tierID,
                tierLabel: profile?.tierLabel
            ))
        case .transport(let message):
            return .failed(message)
        }
    }

    private enum HTTPOutcome {
        case http(Int, Data)
        case transport(String)
    }

    private static func post(url: String, token: String, body: [String: Any]) async -> HTTPOutcome {
        guard let requestURL = URL(string: url),
              let payload = try? JSONSerialization.data(withJSONObject: body) else {
            return .transport("bad request")
        }
        var req = URLRequest(url: requestURL)
        req.httpMethod = "POST"
        req.timeoutInterval = 25 // never let a wedged tunnel hang the poll loop
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.httpBody = payload
        do {
            let (data, response) = try await URLSession.shared.data(for: req)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            return .http(status, data)
        } catch {
            return .transport(error is URLError ? L10n.tr("network drop") : error.localizedDescription)
        }
    }

    // MARK: - Token refresh

    private enum RefreshResult {
        case refreshed(GeminiOAuthCreds)
        case noClientCredentials
        case failed
    }

    /// Exchange the refresh token at Google's token endpoint. Success is
    /// defined as "the writeback landed" — the CLI reads the same file.
    private static func refresh(creds: GeminiOAuthCreds, credsURL: URL) async -> RefreshResult {
        guard let refreshToken = creds.refreshToken else { return .failed }
        guard let client = GeminiClientExtractor.resolve() else { return .noClientCredentials }
        guard let endpoint = URL(string: tokenURL) else { return .failed }

        var req = URLRequest(url: endpoint)
        req.httpMethod = "POST"
        req.timeoutInterval = 20
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.httpBody = formEncoded([
            ("client_id", client.clientID),
            ("client_secret", client.clientSecret),
            ("refresh_token", refreshToken),
            ("grant_type", "refresh_token"),
        ])

        guard let (data, response) = try? await URLSession.shared.data(for: req),
              (response as? HTTPURLResponse)?.statusCode == 200,
              let updated = GeminiCredentials.applyRefreshResponse(data, to: credsURL) else {
            return .failed
        }
        return .refreshed(updated)
    }

    private static func formEncoded(_ pairs: [(String, String)]) -> Data {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        let body = pairs.map { key, value in
            let escaped = value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
            return "\(key)=\(escaped)"
        }.joined(separator: "&")
        return Data(body.utf8)
    }
}

File: windows/src/AgentIsland/Usage/GeminiUsageFetcher.cs
using System.Net.Http;
using System.Text;
using System.Text.Json.Nodes;
using AgentIsland.Localization;

namespace AgentIsland.Usage;

/// Fetches the Gemini CLI's Code Assist quota buckets, authenticating with
/// %USERPROFILE%\.gemini\oauth_creds.json and refreshing through Google's
/// token endpoint when the access token is expired or rejected. Display-only
/// — no transcript monitoring, no alarms live here.
public static class GeminiUsageFetcher
{
    public abstract record Outcome
    {
        public sealed record Success(GeminiQuotaSnapshot Snapshot) : Outcome;

        /// Refresh path exhausted — the user has to run `gemini` and re-auth.
        public sealed record ReauthRequired : Outcome;

        /// A refresh was required but no client id/secret could be extracted
        /// from a local gemini-cli install (and no env override).
        public sealed record NeedsCliInstall : Outcome;

        /// Google's consumer-tier shutdown verdict — the account moved to
        /// Antigravity. A state, not an error.
        public sealed record MigratedToAntigravity : Outcome;

        public sealed record Failed(string Message) : Outcome;

        /// settings.json declares api-key / vertex-ai.
        public sealed record UnsupportedAuth(string AuthType) : Outcome;

        /// No oauth_creds.json — Gemini CLI never signed in on this machine.
        public sealed record NotInstalled : Outcome;
    }

    private const string LoadCodeAssistUrl = "https://cloudcode-pa.googleapis.com/v1internal:loadCodeAssist";
    private const string QuotaUrl = "https://cloudcode-pa.googleapis.com/v1internal:retrieveUserQuota";
    private const string TokenUrl = "https://oauth2.googleapis.com/token";

    /// Per-request ceilings below the shared HttpClient's own timeout: a
    /// wedged tunnel must never hang the poll loop.
    private static readonly TimeSpan QuotaTimeout = TimeSpan.FromSeconds(25);
    private static readonly TimeSpan RefreshTimeout = TimeSpan.FromSeconds(20);

    public static async Task<Outcome> Fetch(CancellationToken ct = default)
    {
        var detection = GeminiCredentials.Detect();
        if (detection is GeminiAuthDetection.UnsupportedAuth unsupported)
        {
            return new Outcome.UnsupportedAuth(unsupported.AuthType);
        }
        if (detection is not GeminiAuthDetection.OauthPersonal) return new Outcome.NotInstalled();

        var credsPath = GeminiCredentials.CredsFile;
        if (GeminiCredentials.LoadCreds(credsPath) is not { } creds) return new Outcome.NotInstalled();

        // Proactive refresh near expiry (Google access tokens live ~1h).
        if (GeminiCredentials.NeedsRefresh(creds) && creds.RefreshToken is not null)
        {
            switch (await Refresh(creds, credsPath, ct))
            {
                case RefreshResult.Refreshed refreshed:
                    creds = refreshed.Creds;
                    break;
                case RefreshResult.NoClientCredentials:
                    return new Outcome.NeedsCliInstall();
                default:
                    // Clock skew can make a working token look expired — still try.
                    break;
            }
        }

        var first = await FetchSnapshot(creds.AccessToken, ct);
        if (first is not SnapshotResult.Unauthorized) return ToOutcome(first);

        if (creds.RefreshToken is null) return new Outcome.ReauthRequired();
        switch (await Refresh(creds, credsPath, ct))
        {
            case RefreshResult.NoClientCredentials:
                return new Outcome.NeedsCliInstall();
            case RefreshResult.Refreshed refreshed:
                return ToOutcome(await FetchSnapshot(refreshed.Creds.AccessToken, ct));
            default:
                return new Outcome.ReauthRequired();
        }
    }

    // MARK: - Quota

    private abstract record SnapshotResult
    {
        public sealed record Success(GeminiQuotaSnapshot Snapshot) : SnapshotResult;
        public sealed record Migrated : SnapshotResult;
        public sealed record Unauthorized : SnapshotResult;
        public sealed record Failed(string Message) : SnapshotResult;
    }

    /// An Unauthorized that reaches here has already spent its retry, so it
    /// resolves to re-auth rather than another refresh round.
    private static Outcome ToOutcome(SnapshotResult result) => result switch
    {
        SnapshotResult.Success success => new Outcome.Success(success.Snapshot),
        SnapshotResult.Migrated => new Outcome.MigratedToAntigravity(),
        SnapshotResult.Failed failed => new Outcome.Failed(failed.Message),
        _ => new Outcome.ReauthRequired(),
    };

    private static async Task<SnapshotResult> FetchSnapshot(string token, CancellationToken ct)
    {
        var profileBody = new JsonObject
        {
            ["metadata"] = new JsonObject
            {
                ["ideType"] = "GEMINI_CLI",
                ["pluginType"] = "GEMINI",
            },
        };

        GeminiQuotaParser.CodeAssistProfile? profile = null;
        switch (await Post(LoadCodeAssistUrl, token, profileBody.ToJsonString(), ct))
        {
            case HttpOutcome.Transport transport:
                return new SnapshotResult.Failed(transport.Message);
            case HttpOutcome.Response { Status: 401 or 403 }:
                return new SnapshotResult.Unauthorized();
            case HttpOutcome.Response response:
                if (GeminiQuotaParser.IsMigrationSignal(response.Body)) return new SnapshotResult.Migrated();
                // The tier/project call is garnish for the quota call — a
                // non-200 here still lets the bucket fetch try with no project.
                if (response.Status == 200) profile = GeminiQuotaParser.ParseLoadCodeAssist(response.Body);
                break;
        }

        var quotaBody = new JsonObject();
        if (profile?.ProjectId is { } projectId) quotaBody["project"] = projectId;
        switch (await Post(QuotaUrl, token, quotaBody.ToJsonString(), ct))
        {
            case HttpOutcome.Transport transport:
                return new SnapshotResult.Failed(transport.Message);
            case HttpOutcome.Response { Status: 401 or 403 }:
                return new SnapshotResult.Unauthorized();
            case HttpOutcome.Response response:
                if (GeminiQuotaParser.IsMigrationSignal(response.Body)) return new SnapshotResult.Migrated();
                if (response.Status != 200) return new SnapshotResult.Failed($"http {response.Status}");
                if (GeminiQuotaParser.ParseQuota(response.Body) is not { } buckets)
                {
                    return new SnapshotResult.Failed("parse error");
                }
                return new SnapshotResult.Success(
                    new GeminiQuotaSnapshot(buckets, profile?.TierId, profile?.TierLabel));
        }

        return new SnapshotResult.Failed("bad response");
    }

    private abstract record HttpOutcome
    {
        public sealed record Response(int Status, byte[] Body) : HttpOutcome;
        public sealed record Transport(string Message) : HttpOutcome;
    }

    private static async Task<HttpOutcome> Post(string url, string token, string json, CancellationToken ct)
    {
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
        timeout.CancelAfter(QuotaTimeout);
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, url);
            request.Headers.TryAddWithoutValidation("Authorization", "Bearer " + token);
            request.Headers.TryAddWithoutValidation("Accept", "application/json");
            request.Content = new StringContent(json, Encoding.UTF8, "application/json");
            using var response = await Http.Client.SendAsync(request, timeout.Token);
            var body = await response.Content.ReadAsByteArrayAsync(timeout.Token);
            return new HttpOutcome.Response((int)response.StatusCode, body);
        }
        catch (Exception error)
        {
            // A dropped connection or a timeout is the common case; the UI
            // shows the shared caption instead of a raw socket message.
            return new HttpOutcome.Transport(
                error is HttpRequestException or OperationCanceledException
                    ? L10n.Tr("network drop")
                    : error.Message);
        }
    }

    // MARK: - Token refresh

    private abstract record RefreshResult
    {
        public sealed record Refreshed(GeminiOAuthCreds Creds) : RefreshResult;
        public sealed record NoClientCredentials : RefreshResult;
        public sealed record Failed : RefreshResult;
    }

    /// Exchange the refresh token at Google's token endpoint. Success is
    /// defined as "the writeback landed" — the CLI reads the same file, and a
    /// rotated token we fail to persist is a token nobody can use again.
    private static async Task<RefreshResult> Refresh(
        GeminiOAuthCreds creds, string credsPath, CancellationToken ct)
    {
        if (creds.RefreshToken is not { } refreshToken) return new RefreshResult.Failed();
        if (GeminiClientExtractor.Resolve() is not { } client) return new RefreshResult.NoClientCredentials();

        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
        timeout.CancelAfter(RefreshTimeout);
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, TokenUrl);
            request.Headers.TryAddWithoutValidation("Accept", "application/json");
            request.Content = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["client_id"] = client.ClientId,
                ["client_secret"] = client.ClientSecret,
                ["refresh_token"] = refreshToken,
                ["grant_type"] = "refresh_token",
            });
            using var response = await Http.Client.SendAsync(request, timeout.Token);
            if ((int)response.StatusCode != 200) return new RefreshResult.Failed();
            var body = await response.Content.ReadAsByteArrayAsync(timeout.Token);
            if (GeminiCredentials.ApplyRefreshResponse(body, credsPath) is not { } updated)
            {
                return new RefreshResult.Failed();
            }
            return new RefreshResult.Refreshed(updated);
        }
        catch
        {
            return new RefreshResult.Failed();
        }
    }
}
